src/drv_vrtek/vrtek.c

```diff
@@ -0,0 +1,548 @@
+    }
+}
+
+static int getf(ohmd_device* device, ohmd_float_value type, float* out)
+{
+    vrtek_priv* priv = vrtek_priv_get(device);
+
+    switch (type) {
+    case OHMD_ROTATION_QUAT:
+        *(quatf*)out = *(quatf*)&priv->hmd_data.quaternion;
+        break;
+
+    case OHMD_POSITION_VECTOR:
+        out[0] = out[1] = out[2] = 0;
+        break;
+
+    case OHMD_DISTORTION_K:
+        /* FIXME: update this with real values */
+        memset(out, 0, sizeof(float) * 6);
+        break;
+
+    default:
+        ohmd_set_error(priv->device.ctx, "invalid type given to getf (%ud)",
+                       type);
+        return -1;
+        break;
+    }
+
+    return 0;
+}
+
+static void close_device(ohmd_device* device)
+{
+    LOGD("closing device");
+    vrtek_priv* priv = vrtek_priv_get(device);
+    hid_close(priv->hid_handle);
+    free(priv);
+}
+
+#define UDEV_WIKI_URL "https://github.com/OpenHMD/OpenHMD/wiki/Udev-rules-list"
+static ohmd_device* open_device(ohmd_driver* driver, ohmd_device_desc* desc)
+{
+    vrtek_priv* priv = ohmd_alloc(driver->ctx, sizeof(vrtek_priv));
+    if (!priv)
+        goto cleanup;
+
+    priv->device.ctx = driver->ctx;
+
+    /* Open the HID device */
+    priv->hid_handle = hid_open_path(desc->path);
+
+    if (!priv->hid_handle) {
+        char* path = _hid_to_unix_path(desc->path);
+        ohmd_set_error(driver->ctx, "Could not open %s.\n"
+                       "Check your permissions: "
+                       UDEV_WIKI_URL, path);
+        free(path);
+        goto cleanup;
+    }
+
+    if (hid_set_nonblocking(priv->hid_handle, 1) == -1) {
+        ohmd_set_error(driver->ctx, "failed to set non-blocking on device");
+        goto cleanup;
+    }
+
+    /* Set default device properties */
+    ohmd_set_default_device_properties(&priv->device.properties);
+
+    vrtek_send_init(priv);
+
+    /* Disable IMU */
+    vrtek_set_imu_state(priv, false);
+
+    /* Read the config */
+    vrtek_read_config(priv);
+
+    /* Read the model */
+    vrtek_read_model(priv);
+
+    /* Enable IMU */
+    vrtek_set_imu_state(priv, true);
+
+    priv->device.update = update_device;
+    priv->device.close = close_device;
+    priv->device.getf = getf;
+    priv->device.settings.automatic_update = 0;
+
+    /* calculate eye projection matrices from the device properties */
+    ohmd_calc_default_proj_matrices(&priv->device.properties);
+
+    return &priv->device;
+
+cleanup:
+    if (priv)
+        free(priv);
+
+    return NULL;
+}
+
+static void get_device_list(ohmd_driver* driver, ohmd_device_list* list)
+{
+    /* Enumerate HID devices and add any VR-Tek HMDs found to the device list */
+    struct hid_device_info* devs = hid_enumerate(VRTEK_ID, VRTEK_WVR_HMD);
+    struct hid_device_info* cur_dev = devs;
+
+    while (cur_dev) {
+        /* This is needed because VR-Tek uses the Oculus DK1 USB ID */
+        if (wcscmp(cur_dev->manufacturer_string, L"STMicroelectronics")==0 &&
```

Comment:
Quality firmware... Is there anything that actually identifies these as VR-Tek?

Reply:
This is similar to some other devices we have (such as the Deepoon E2)

Reply:
> Quality firmware... Is there anything that actually identifies these as VR-Tek?
Unfortunately there's not at this level :(

> This is similar to some other devices we have (such as the Deepoon E2)
Yeah, thankfully I saw that NOLO/Deepoon example. I was a little worried about what to do with this looking like a DK1 until I noticed that.

Reply:
I see, for now we'll have to assume all DK1 vid/pid devices with "STMicroelectronics" / "HID" manufacturer / product strings are VR-Tek.

Reply:
That makes sense. Sounds good. Thanks!